src/python/pants/testutil/rule_runner.py

```diff
@@ -656,89 +659,163 @@ class MockGet(Generic[_O]):
 
 @dataclass(frozen=True)
 class MockRequestExceptionComparable:
-    category: Literal["Get"] | Literal["Effect"] | str
-    output_type: str | None
-    input_types: tuple[str, ...]
+    category: Literal["Get", "Effect"]
+    output_type: type | None
+    input_types: tuple[type, ...]
 
     def __str__(self):
-        inputs = ",".join(str(e) for e in self.input_types)
-        return f"{self.category}({self.output_type}, ({inputs},))"
+        inputs = ", ".join(e.__name__ for e in self.input_types)
+        return f"{self.category}({self.output_type.__name__}, ({inputs}))"
 
 
 def _compare_expected_mocks(
-    expected: Sequence[Get | Effect | AwaitableConstraints],
-    actual: Sequence[MockGet | MockEffect],
-) -> str:
-    """Try to be helpful with identifying the problem with supplied mocks."""
+    expected: Sequence[AwaitableConstraints],
+    actual_calls: Mapping[str, Callable[..., _O]],
+    actual_gets: Sequence[MockGet | MockEffect],
+    union_membership: UnionMembership | None,
```

Comment:
This validation function never took union membership into account, and was always broken. We never noticed because in practice it rarely ran (see below).